ethers-providers/src/provider.rs

```diff
@@ -786,6 +786,151 @@ impl<P: JsonRpcClient> Middleware for Provider<P> {
+    async fn resolve_avatar(&self, ens_name: &str) -> Result<Url, ProviderError> {
+        let field = self.resolve_field(ens_name, "avatar").await?;
+        let url = Url::from_str(&field).map_err(|e| ProviderError::CustomError(e.to_string()))?;
+        let owner = self.resolve_name(ens_name).await?;
+        match url.scheme() {
+            "https" | "data" => Ok(url),
+            "ipfs" => erc::http_link_ipfs(url).map_err(ProviderError::CustomError),
+            "eip155" => {
+                let split: Vec<&str> = url.path().trim_start_matches("1/").split(':').collect();
+                let (inner_scheme, inner_path) = if split.len() == 2 {
+                    (split[0], split[1])
+                } else {
+                    return Err(ProviderError::CustomError("Unsupported ERC link".to_string()));
+                };
+
+                let token_split: Vec<&str> = inner_path.split('/').collect();
+                let (contract_addr, token_id) = if token_split.len() == 2 {
+                    let token_id = U256::from_dec_str(token_split[1]).map_err(|e| {
+                        ProviderError::CustomError(format!(
+                            "Unsupported token id type: {} {}",
+                            token_split[1], e
+                        ))
+                    })?;
+                    let mut token_id_bytes = [0x0; 32];
+                    token_id.to_big_endian(&mut token_id_bytes);
+                    (
+                        Address::from_str(token_split[0].trim_start_matches("0x")).map_err(
+                            |e| {
+                                ProviderError::CustomError(format!(
+                                    "Invalid contract address: {} {}",
+                                    token_split[0], e
+                                ))
+                            },
+                        )?,
+                        token_id_bytes,
+                    )
+                } else {
+                    return Err(ProviderError::CustomError(
+                        "Unsupported ERC link path".to_string(),
+                    ));
+                };
+                let selector: Selector = match inner_scheme {
+                    "erc721" => {
+                        let tx = TransactionRequest {
+                            data: Some([&erc::ERC721_SELECTOR[..], &token_id].concat().into()),
+                            to: Some(NameOrAddress::Address(contract_addr)),
+                            ..Default::default()
+                        };
+                        let data = self.call(&tx.into(), None).await?;
+                        if decode_bytes::<Address>(ParamType::Address, data) != owner {
+                            return Err(ProviderError::CustomError("Incorrect owner.".to_string()));
+                        }
+                        [0xc8, 0x7b, 0x56, 0xdd]
+                    }
+                    "erc1155" => {
```

Comment:
if possible, we should make them 2 separate functions

Reply:
307bdcc adds a `ERCToken` type and another method to resolve a token. Not sure about how future this is though :thinking: